receiver/prometheusexecreceiver/subprocessmanager/manager.go

```diff
@@ -29,33 +29,33 @@ import (
 	"go.uber.org/zap"
 )
 
-// Process struct holds all the info needed to instantiate a subprocess
-type Process struct {
-	Command string
-	Port    int
-	Env     []EnvConfig
-}
-
 const (
-	// HealthyProcessTime is the default time a process needs to stay alive to be considered healthy
-	HealthyProcessTime time.Duration = 30 * time.Minute
-	// HealthyCrashCount is the amount of times a process can crash (within the healthyProcessTime) before being considered unstable - it may be trying to find a port
-	HealthyCrashCount int = 3
 	// delayMutiplier is the factor by which the delay scales
 	delayMultiplier float64 = 2.0
 	// initialDelay is the initial delay before a process is restarted
 	initialDelay time.Duration = 1 * time.Second
 )
 
+// GetDelay will compute the delay for a given process according to its crash count and time alive using an exponential backoff algorithm
+func GetDelay(elapsed time.Duration, healthyProcessDuration time.Duration, crashCount int, healthyCrashCount int) time.Duration {
```

Comment:
This function is not really connected to the rest of this package, and seems to be _very tightly coupled_ with logic in your receiver. How about bringing it up to the receiver?

Reply:
I can bring it to the receiver, but I do think it should be in the subprocess manager. Even if I use it inside the receiver, the logic would apply to any subprocess manager trying to compute the delay of restarting a process. Processes shouldn't have exponential backoff applied unless they're having an issue (crashing before a reasonable amount of time), they should be restarted as the base delay if so. Also `crashCount` may be a `prometheus_exec` specific thing but I think it provides flexibility, and can be set to 1 if need be as an argument. Let me know!

Reply:
Generally speaking I agree that this kind of logic, that relates to running / restarting processes, should be in the package that manages processes. Ideally, I would prefer if more logic was moved into the process manager package, and the package interface updated to provide hooks to update the command on restart or something along those lines as discussed in https://github.com/open-telemetry/opentelemetry-collector/pull/1483#issuecomment-668943861), so the prom exec logic can be nicely decoupled from the process management work.

But at the moment you have most of the logic related to restarting processes currently in the receiver. Due to this high level of coupling between that package and thus function, I think you are better keeping this function in the receiver for now.

Reply:
Alright. sounds good!